### ISSUE TITLE
Handle simulatneous connections

### ISSUE THREAD

**Issue by Tiino1** [kitswas/VirtualGamePad-PC#8]
This project would be a great replacement to DroidJoy if we could use it with multiple android phones (clients).

This issue asks:
- either to make 1 server handle multiple clients
- either to allow multiple servers to run in parallel, each allowing 1 clients
-- in that case each clients must have a dedicated emulated device on the host PC to avoid interferences between commands

Finally as an enhancement it would be great to :
- handle an undetermined number of players
-- unlike DroidJoy which supports up to (a fixed number of) 4 players

**Comment by kitswas** [comment]
Can you be more detailed?

From [the server's documentation](https://kitswas.github.io/VirtualGamePad-PC/index.html#how-it-works) page,

> 1. A socket connection is established between the client and the server.
>
> 2. The server uses [the SendInput() function](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-sendinput) to synthesize keystrokes, mouse motions, and button clicks depending on the input from the client.  
> It has the following implications:
>    - The virtual gamepad is not limited to games. You can use the virtual gamepad to control any application on the computer.
>    - Games that use DirectInput or XInput do not recognize input from the virtual gamepad as gamepad input but as keyboard and mouse input. If required, the controls have to be remapped in the game.
>
> 3. The buttons on the virtual gamepad are mapped to the keyboard keys and mouse actions using the maps at src/networking/executor.hpp.  

**TL;DR**: The server calls the Windows API for input injection, it does not create virtual input devices.

For virtual input devices, here's what I know:

1. Use [Windows.UI.Input.Preview.Injection](https://learn.microsoft.com/en-us/windows/apps/design/input/input-injection). It [does not work](https://kitswas.github.io/VirtualGamePad-PC/index.html#history-of-the-servers-development).
2. Create your own virtual device and drivers. (Not feasible)  
3. Use a library that does 2. [ViGEmBus])(https://github.com/nefarius/ViGEmBus) was great until it got [discontinued](https://docs.nefarius.at/projects/ViGEm/End-of-Life/).

Are you aware of any other methods?

**Comment by kitswas** [comment]
Feel free to continue the discussion.  
The issue can be reopened if a potential way to implement this is found.  
Currently, this is out of scope till the basics are done and tested.  
The software is still pre-release.

I appreciate your feedback. The knowledge that people are interested keeps me up at night working on projects like these. 😃

**Comment by kitswas** [comment]
@Tiino1 

[v0.4.0](https://github.com/kitswas/VirtualGamePad-PC/releases/tag/v0.4.0) was released yesterday.

The Linux version of the server can have multiple instances running in parallel, each allowing a single client.

On Windows, the _**experimental** GamepadExecutor_ can do the same. (Note that it is experimental and sometimes crashes on launch. Try launching again till it works.)
